core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -77,17 +77,44 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
 
   private[spark] var scheduler: TaskScheduler = null
 
-  // executor ID -> timestamp of when the last heartbeat from this executor was received
+  /**
+   * [SPARK-39984]
+   * Please make sure the intersection between `executorLastSeen` and `waitingList` is an empty set.
+   * If the intersection is not empty, it is possible to never kill the executor until the executor
+   * recovers. When an executor is in both `executorLastSeen` and `waitingList`, the value of
+   * `workerLastHeartbeat` in waitingList may update if the worker sends heartbeats to master
+   * normally.
+   *
+   * `executorLastSeen`:
+   *  - key: executor ID
+   *  - value: timestamp of when the last heartbeat from this executor was received
+   *
+   *  `waitingList`: executor ID -> WorkerLastHeartbeat
+   *  - key: executor ID
+   *  - value: timestamp of when the last heartbeat from the worker was received
+   *
+   * when driver does not receive any heartbeat from an executor for `executorTimeoutMs` seconds,
+   * the driver will ask master for the last heartbeat from the worker which the executor is running
+   * on.
+   */
   private val executorLastSeen = new HashMap[String, Long]
+  private val waitingList = new HashMap[String, Long]
 
   private val executorTimeoutMs = sc.conf.get(
     config.STORAGE_BLOCKMANAGER_HEARTBEAT_TIMEOUT
-  ).getOrElse(Utils.timeStringAsMs(s"${sc.conf.get(Network.NETWORK_TIMEOUT)}s"))
+  ).getOrElse(Utils.timeStringAsMs(s"${sc.conf.get(Network.NETWORK_EXECUTOR_TIMEOUT)}s"))
 
   private val checkTimeoutIntervalMs = sc.conf.get(Network.NETWORK_TIMEOUT_INTERVAL)
 
   private val executorHeartbeatIntervalMs = sc.conf.get(config.EXECUTOR_HEARTBEAT_INTERVAL)
 
+  /**
+   * Currently, [SPARK-39984] is only for StandaloneSchedulerBackend.
+   */
+  private val checkWorkerLastHeartbeat =
+    sc.conf.get(HEARTBEAT_RECEIVER_CHECK_WORKER_LAST_HEARTBEAT) &&
+      sc.schedulerBackend.isInstanceOf[StandaloneSchedulerBackend]
```

Comment:
When `HeartbeatReceiver` is created in `SparkContext`, `_schedulerBackend` is not yet created/initialized - so we cannot depend on this here (`sc.schedulerBackend` will always return `null`)

Reply:
Thank you for pointing this out! I have fixed this in https://github.com/apache/spark/pull/37411/commits/9afdcfa5b71d2970ed7518f770aedadd3ccb82f1.

Reply:
TODO: I will check whether SchedulerBackend is initialized or not before HeartbeatReceiver receives `ExecutorRegistered` / `ExecutorRemoved` / `Heartbeat`.

Reply:
Heartbeat receiver will receive all of the messages (`ExecutorRegistered` / `ExecutorRemoved` / `Heartbeat`) only after SchedulerBackend is initialized. Take `ExecutorRegistered` as an example, scheduler backend has already been initialized in Step1, and used the message `SparkListenerExecutorAdded` to notify HeartbeatReceiver to register the new executor.

* Step1: SchedulerBackend (CoarseGrainedSchedulerBackend / LocalSchedulerBackend)
   * `post(SparkListenerExecutorAdded)`
* Step2: `doPostEvent` (SparkListenerBus) 
* Step3: `onExecutorAdded` (HeartbeatReceiver)
* Step4: `addExecutor` (HeartbeatReceiver)
* Step5: `ExecutorRegistered` (HeartbeatReceiver)

Reply:
I have revisited all `sc.schedulerBackend` occurrences to prevent the exceptions caused by the lack of scheduler backend.


1. `ExecutorRegistered` / `ExecutorRemoved` / `Heartbeat` => As mentioned above, heartbeat receiver will receive all of these messages only after SchedulerBackend is initialized.

2. `killExecutor` => pattern matching is exhaustive

3. `expireDeadHosts` => Handled by new commit https://github.com/apache/spark/pull/37411/commits/77192300cb6990dfc60acc3227c6ae95e0223726

Reply:
Note that you can actually have `ExecutorRegistered` before `sc.schedulerBackend` is set - this is essentially a race condition. This is why we have a `if (scheduler != null) {` check while handling `Heartbeat`

Reply:
Hi @mridulm, the event `SparkListenerExecutorAdded` will only be posted by `LocalSchedulerBackend` ([Link](https://sourcegraph.com/github.com/apache/spark@b012cb722f6105a945e971bdc509f803a90b7419/-/blob/core/src/main/scala/org/apache/spark/scheduler/local/LocalSchedulerBackend.scala?L134)), `CoarseGrainedSchedulerBackend` ([Link](https://sourcegraph.com/github.com/apache/spark@b012cb722f6105a945e971bdc509f803a90b7419/-/blob/core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala?L299:13)), and `MesosFineGrainedSchedulerBackend` ([Link](https://sourcegraph.com/github.com/apache/spark@b012cb722f6105a945e971bdc509f803a90b7419/-/blob/resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosFineGrainedSchedulerBackend.scala?L332:28)). Hence, I think `ExecutorRegistered` will always happen after `sc.schedulerBackend` initialization. 

Do I miss any context? Thank you!

Reply:
This code has evolved a lot, so it is possible that the check in handling `Heartbeat` is no longer valid (the `null` check there is equally not likely given your comment above) - having said that, I have not looked at all corner cases, so would be careful.

Reply:
@Ngone51 Here is the other discussion about the lack of scheduler backend.